Python/specialize.c

```diff
@@ -1138,14 +1138,36 @@ binary_subscr_faiL_kind(PyTypeObject *container_type, PyObject *sub)
 }
 #endif
 
+_Py_IDENTIFIER(__getitem__);
+
+static int
+function_kind(PyCodeObject *code) {
+    int flags = code->co_flags;
+    if (flags & (CO_GENERATOR | CO_COROUTINE | CO_ASYNC_GENERATOR)) {
+        return SPEC_FAIL_GENERATOR;
+        return -1;
+    }
+    if ((flags & (CO_VARKEYWORDS | CO_VARARGS)) || code->co_kwonlyargcount) {
+        return SPEC_FAIL_COMPLEX_PARAMETERS;
+    }
+    if ((flags & CO_OPTIMIZED) == 0) {
+        return SPEC_FAIL_CO_NOT_OPTIMIZED;
+    }
+    if (code->co_nfreevars) {
+        return SPEC_FAIL_FREE_VARS;
+    }
+    return 0;
```

Comment:
I don't really like this, since `0` is a valid failure code. Perhaps return `-1` on success and check for that instead (it's sort of a weird interface, but whatever).

Reply:
I'll add a success code, for clarity.

In many other C programs, yes `0` is a failure.
However, in CPython, `0` is usually a success, as `-1` (or any negative number in some cases) is a failure.
E.g. `PyObject_RichCompareBool` return `0` for False, and `-1` for an error.

`compile.c` is an unfortunate counter example, where some functions return `0` for a failure and others return `0` for a success.